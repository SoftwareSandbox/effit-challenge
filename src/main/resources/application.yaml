ktor:
  application:
    modules:
      - be.swsb.effit.ApplicationKt.module
  deployment:
    port: 8080
oauth:
  google:
    clientId: "239901082410-csmlhg27d90ela85l8pmk9gt4gikrkj3.apps.googleusercontent.com"
    clientSecret: "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"