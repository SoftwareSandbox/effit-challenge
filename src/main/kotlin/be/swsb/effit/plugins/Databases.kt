package be.swsb.effit.plugins

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import java.sql.*
import kotlinx.coroutines.*

fun Application.configureDatabases() {
    val dbConnection: Connection = connectToPostgres(embedded = true)
    val cityService = CityService(dbConnection)
    routing {
        // Create city
        post("/cities") {
            val city = call.receive<City>()
            val id = cityService.create(city)
            call.respond(HttpStatusCode.Created, id)
        }
        // Read city
        get("/cities/{id}") {
            val id = call.parameters["id"]?.toInt() ?: throw IllegalArgumentException("Invalid ID")
            try {
                val city = cityService.read(id)
                call.respond(HttpStatusCode.OK, city)
            } catch (e: Exception) {
                call.respond(HttpStatusCode.NotFound)
            }
        }
        // Update city
        put("/cities/{id}") {
            val id = call.parameters["id"]?.toInt() ?: throw IllegalArgumentException("Invalid ID")
            val user = call.receive<City>()
            cityService.update(id, user)
            call.respond(HttpStatusCode.OK)
        }
        // Delete city
        delete("/cities/{id}") {
            val id = call.parameters["id"]?.toInt() ?: throw IllegalArgumentException("Invalid ID")
            cityService.delete(id)
            call.respond(HttpStatusCode.OK)
        }
    }
}

/**
 * Makes a connection to a Postgres database.
 *
 * In order to connect to your running Postgres process,
 * please specify the following parameters in your configuration file:
 * - postgres.url -- Url of your running database process.
 * - postgres.user -- Username for database connection
 * - postgres.password -- Password for database connection
 *
 * If you don't have a database process running yet, you may need to [download]((https://www.postgresql.org/download/))
 * and install Postgres and follow the instructions [here](https://postgresapp.com/).
 * Then, you would be able to edit your url,  which is usually "jdbc:postgresql://host:port/database", as well as
 * user and password values.
 *
 *
 * @param embedded -- if [true] defaults to an embedded database for tests that runs locally in the same process.
 * In this case you don't have to provide any parameters in configuration file, and you don't have to run a process.
 *
 * @return [Connection] that represent connection to the database. Please, don't forget to close this connection when
 * your application shuts down by calling [Connection.close]
 * */
fun Application.connectToPostgres(embedded: Boolean): Connection {
    Class.forName("org.postgresql.Driver")
    if (embedded) {
        return DriverManager.getConnection("jdbc:h2:mem:test;DB_CLOSE_DELAY=-1", "root", "")
    } else {
        val url = environment.config.property("postgres.url").getString()
        val user = environment.config.property("postgres.user").getString()
        val password = environment.config.property("postgres.password").getString()

        return DriverManager.getConnection(url, user, password)
    }
}
