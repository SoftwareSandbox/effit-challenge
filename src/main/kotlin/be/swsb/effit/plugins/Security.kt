package be.swsb.effit.plugins

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.http.headers
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.config.*
import io.ktor.server.html.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.sessions.*
import kotlinx.html.body
import kotlinx.html.div
import kotlinx.html.img
import kotlinx.html.p
import kotlinx.serialization.Serializable

fun Application.configureSecurity(httpClient: HttpClient) {

    val redirects = mutableMapOf<String, String>()

    val googleConfig = environment.config.config("oauth").config("google")
    val googleClientId = googleConfig.tryGetString("clientId") ?: "clientId"
    val googleClientSecret = googleConfig.tryGetString("clientSecret") ?: "clientSecret"
    authentication {
        oauth("auth-oauth-google") {
            urlProvider = { "http://localhost:8080/callback" }
            providerLookup = {
                OAuthServerSettings.OAuth2ServerSettings(
                    name = "google",
                    authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
                    accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
                    requestMethod = HttpMethod.Post,
                    clientId = googleClientId,
                    clientSecret = googleClientSecret,
                    defaultScopes = listOf("https://www.googleapis.com/auth/userinfo.profile"),
                    extraAuthParameters = listOf("access_type" to "offline"),
                    onStateCreated = { call, state ->
                        redirects[state] =
                            call.request.queryParameters["redirectUrl"] ?: error("No redirectUrl query param")
                    }
                )
            }
            client = httpClient
        }
    }
    install(Sessions) {
        cookie<UserSession>("MY_SESSION") {
            cookie.extensions["SameSite"] = "lax"
        }
    }
    routing {
        authenticate("auth-oauth-google") {
            get("login") {
                // Redirects to 'authorizeUrl' automatically
            }

            get("/callback") {
                val principal: OAuthAccessTokenResponse.OAuth2? = call.principal()
                call.sessions.set(UserSession(principal!!.state!!, principal.accessToken))
                val redirect = redirects[principal.state!!]
                call.respondRedirect(redirect!!)
            }
        }
        get("/{path}") {
            val userSession: UserSession? = call.sessions.get()
            if (userSession != null) {
                val response: HttpResponse = httpClient.get("https://www.googleapis.com/oauth2/v2/userinfo") {
                    headers {
                        append(HttpHeaders.Authorization, "Bearer ${userSession.token}")
                    }
                }
                println(response.bodyAsText())
                val userInfo: UserInfo = response.body()
                call.respondHtml {
                    body {
                        div {
                            p { +"Hello, ${userInfo.name}!" }
                            img { src = userInfo.picture }
                        }
                    }
                }
            } else {
                val redirectUrl = URLBuilder("http://0.0.0.0:8080/login").run {
                    parameters.append("redirectUrl", call.request.uri)
                    build()
                }
                call.respondRedirect(redirectUrl)
            }
        }
    }
}

data class UserSession(val state: String, val token: String)

@Serializable
data class UserInfo(
    val id: String,
    val name: String,
    val picture: String,
    val locale: String,
)