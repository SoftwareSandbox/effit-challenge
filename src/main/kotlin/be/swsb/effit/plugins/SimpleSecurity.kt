package be.swsb.effit.plugins

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.http.HttpStatusCode.Companion.Unauthorized
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.config.*
import io.ktor.server.html.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.sessions.*
import io.ktor.util.pipeline.*
import kotlinx.html.*
import kotlinx.serialization.Serializable

fun Application.configureSimpleSecurity(httpClient: HttpClient) {

    val redirects = mutableMapOf<String, String>()

    val googleConfig = environment.config.config("oauth").config("google")
    val keycloakConfig = environment.config.config("oauth").config("keycloak")
    val googleClientId = googleConfig.tryGetString("clientId") ?: "clientId"
    val googleClientSecret = googleConfig.tryGetString("clientSecret") ?: "clientSecret"
    val keycloakClientId = keycloakConfig.tryGetString("clientId") ?: "clientId"
    val keycloakClientSecret = keycloakConfig.tryGetString("clientSecret") ?: "clientSecret"

    authentication {
//        google(googleClientId, googleClientSecret, redirects, httpClient)
        keycloak(keycloakClientId, keycloakClientSecret, redirects, httpClient)
    }

    install(Sessions) {
        cookie<UserSession>("MY_SESSION") {
            cookie.extensions["SameSite"] = "lax"
        }
    }

    routing {
        authenticate("auth-oauth-keycloak") {
            get("login") {
                // Redirects to 'authorizeUrl' automatically
            }

            get("/callback") {
                val principal: OAuthAccessTokenResponse.OAuth2? = call.principal()
                call.sessions.set(UserSession(principal!!.state!!, principal.accessToken))
                val redirect = redirects[principal.state!!]
                call.respondRedirect(redirect!!)
            }
        }
        get("/{path}") {
            val userSession: UserSession? = call.sessions.get()
            if (userSession != null) {
                val userInfo: UserInfo = httpClient.get("https://www.googleapis.com/oauth2/v2/userinfo") {
                    headers {
                        append(HttpHeaders.Authorization, "Bearer ${userSession.token}")
                    }
                }.body()
                call.respondText("Hello, ${userInfo.name}!")
            } else {
                val redirectUrl = URLBuilder("http://0.0.0.0:8080/login").run {
                    parameters.append("redirectUrl", call.request.uri)
                    build()
                }
                call.respondRedirect(redirectUrl)
            }
        }
    }
}

private fun AuthenticationConfig.keycloak(
    clientId: String,
    clientSecret: String,
    redirects: MutableMap<String, String>,
    httpClient: HttpClient
) {
    oauth("auth-oauth-keycloak") {
        urlProvider = { "http://localhost:8080/callback" }
        providerLookup = {
            OAuthServerSettings.OAuth2ServerSettings(
                name = "keycloak",
                authorizeUrl = "http://localhost:9090/realms/effit-ktor/protocol/openid-connect/auth",
                accessTokenUrl = "http://localhost:9090/realms/effit-ktor/protocol/openid-connect/token",
                requestMethod = HttpMethod.Post,
                clientId = clientId,
                clientSecret = clientSecret,
                defaultScopes = listOf("profile"),
                extraAuthParameters = listOf("access_type" to "offline"),
                onStateCreated = { call, state ->
                    redirects[state] =
                        call.request.queryParameters["redirectUrl"] ?: error("No redirectUrl query param")
                },
            )
        }
        client = httpClient
    }
}
private fun AuthenticationConfig.google(
    googleClientId: String,
    googleClientSecret: String,
    redirects: MutableMap<String, String>,
    httpClient: HttpClient
) {
    oauth("auth-oauth-google") {
        urlProvider = { "http://localhost:8080/callback" }
        providerLookup = {
            OAuthServerSettings.OAuth2ServerSettings(
                name = "google",
                authorizeUrl = "https://accounts.google.com/o/oauth2/auth",
                accessTokenUrl = "https://accounts.google.com/o/oauth2/token",
                requestMethod = HttpMethod.Post,
                clientId = googleClientId,
                clientSecret = googleClientSecret,
                defaultScopes = listOf("https://www.googleapis.com/auth/userinfo.profile"),
                extraAuthParameters = listOf("access_type" to "offline"),
                onStateCreated = { call, state ->
                    redirects[state] =
                        call.request.queryParameters["redirectUrl"] ?: error("No redirectUrl query param")
                },
            )
        }
        client = httpClient
    }
}

data class UserSession2(val state: String, val token: String)

@Serializable
data class UserInfo2(
    val id: String,
    val name: String,
    val picture: String,
    val locale: String,
)