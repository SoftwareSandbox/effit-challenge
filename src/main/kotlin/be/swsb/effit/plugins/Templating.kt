package be.swsb.effit.plugins

import io.ktor.server.application.*
import io.ktor.server.html.*
import io.ktor.server.routing.*
import kotlinx.html.*

fun Application.configureTemplating() {
    routing {
        get("/html-dsl") {
            call.respondHtml {
                body {
                    h1 { +"HTML" }
                    ul {
                        for (n in 1..10) {
                            li { +"$n" }
                        }
                    }
                }
            }
        }
    }
}
