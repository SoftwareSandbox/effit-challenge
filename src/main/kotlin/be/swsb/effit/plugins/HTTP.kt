package be.swsb.effit.plugins

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.plugins.compression.*
import io.ktor.server.plugins.cors.routing.*
import io.ktor.server.plugins.forwardedheaders.*

fun Application.configureHTTP() {
    install(Compression) {
        gzip {
            priority = 1.0
        }
        deflate {
            priority = 10.0
            minimumSize(1024) // condition
        }
    }
    install(CORS) {
        allowMethod(HttpMethod.Options)
        allowMethod(HttpMethod.Put)
        allowMethod(HttpMethod.Delete)
        allowMethod(HttpMethod.Patch)
        allowHeader(HttpHeaders.Authorization)
//        allowHeader(HttpHeaders.AccessControlAllowOrigin)
//        allowHost("accounts.google.com",listOf("https"))
        allowHeader("MyCustomHeader")
        anyHost() // @TODO: Don't do this in production if possible. Try to limit it.
    }
    install(ForwardedHeaders) // WARNING: for security, do not include this if not behind a reverse proxy
    install(XForwardedHeaders) // WARNING: for security, do not include this if not behind a reverse proxy
}
