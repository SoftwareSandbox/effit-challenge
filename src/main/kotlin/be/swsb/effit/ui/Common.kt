package be.swsb.effit.ui

import io.ktor.http.*
import io.ktor.server.plugins.*
import io.ktor.server.util.*

interface Command {
    val url: String
}

interface Query {
    val url: String
}

inline fun <reified R : Any> Parameters.getOrNull(name: String): R? = try {
    getOrFail<R>(name)
} catch (e: MissingRequestParameterException) {
    null
}