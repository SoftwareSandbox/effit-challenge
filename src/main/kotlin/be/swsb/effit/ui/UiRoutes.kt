package be.swsb.effit.ui

import be.swsb.effit.plugins.withUser
import be.swsb.effit.ui.pages.EffitUi
import be.swsb.effit.ui.views.AboutView
import be.swsb.effit.ui.views.HomeView
import be.swsb.effit.ui.views.NavigableView
import be.swsb.effit.ui.views.ProfileView
import io.ktor.client.*
import io.ktor.server.application.*
import io.ktor.server.routing.*

fun Application.configureUI(httpClient: HttpClient) {
    routing {
        uiRoutes(httpClient)
    }
}

fun Routing.uiRoutes(httpClient: HttpClient) {
    fun Routing.registerGetFor(view: NavigableView) {
        route(view.url) {
            get {
                withUser(httpClient) { userInfo ->
                    EffitUi(navigableView = view, userInfo = userInfo)
                }
            }
        }
    }

    val navigableViews = listOf(HomeView, AboutView, ProfileView)
    navigableViews.forEach(::registerGetFor)
}

