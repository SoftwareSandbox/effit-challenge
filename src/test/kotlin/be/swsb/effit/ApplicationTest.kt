package be.swsb.effit

import be.swsb.effit.plugins.UserInfo
import be.swsb.effit.plugins.UserSession
import be.swsb.effit.plugins.configureSecurity
import be.swsb.effit.ui.configureUI
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.cookies.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.server.application.*
import io.ktor.server.config.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.sessions.*
import io.ktor.server.testing.*
import org.approvaltests.Approvals
import kotlin.test.Test

class ApplicationTest {
    @Test
    fun testSecurity() = testApplication {
        environment {
            config = ApplicationConfig("application-custom.yaml")
        }
        val testHttpClient = createClient {
            install(HttpCookies)
            install(ContentNegotiation) {
                json()
            }
        }
        application {
            configureSecurity(testHttpClient)
            configureUI(testHttpClient)
        }

        routing {
            get("/login-test") {
                call.sessions.set(UserSession("xyzABC123","abc123"))
            }
        }

        externalServices {
            hosts("https://www.googleapis.com") {
                install(io.ktor.server.plugins.contentnegotiation.ContentNegotiation) {
                    json()
                }
                routing {
                    get("oauth2/v2/userinfo") {
                        call.respond(UserInfo("1", "JetBrains", "image-url", ""))
                    }
                }
            }
        }
        val loginResponse = testHttpClient.get("/login-test")
        val helloResponse = testHttpClient.get("/")

        Approvals.verify(helloResponse.bodyAsText())
    }
}
