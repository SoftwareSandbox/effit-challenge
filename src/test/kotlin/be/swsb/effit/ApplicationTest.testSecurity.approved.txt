<!DOCTYPE html>
<html>
  <head>
    <title>Effit Challenge</title>
    <meta charset="UTF-8">
    <link href="https://cdn.jsdelivr.net/npm/bulma@0.9.4/css/bulma.min.css" rel="stylesheet" type="text/css">
    <link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.0/css/all.min.css" rel="stylesheet" type="text/css">
    <link href="/static/favicon-16x16.png" rel="icon" type="image/png" sizes="16x16">
    <link href="/static/favicon-32x32.png" rel="icon" type="image/png" sizes="32x32">
    <script src="https://unpkg.com/htmx.org@1.8.4" integrity="sha384-wg5Y/JwF7VxGk4zLsJEcAojRtlVp1FKKdGy1qN+OMtdq72WRvX/EdRdqg/LOhYeV" crossorigin="anonymous"></script>
    <script src="https://unpkg.com/htmx.org/dist/ext/multi-swap.js"></script>
  </head>
  <body hx-ext="multi-swap">
    <div class="is-flex is-flex-direction-column" id="EffitUi">
      <nav class="navbar is-primary">
        <div class="navbar-menu is-size-5">
          <div class="navbar-start" hx-boost="true"><a href="/" class="navbar-item is-size-5">Home</a>
            <div class="navbar-item has-dropdown is-hoverable"><a class="navbar-link">More</a>
              <div class="navbar-dropdown" hx-boost="true"><a href="/about" class="navbar-item is-size-5">About</a>
                <hr class="navbar-divider">
<a href="/" class="navbar-item is-size-5">Report an issue</a></div>
            </div>
          </div>
          <div class="navbar-end">
            <div class="navbar-item">
              <div class="media">
                <div class="media-left">
                  <figure class="image is-48x48"><img alt="Placeholder image" src="image-url"></figure>
                </div>
                <div class="media-content">
                  <p class="is-4">JetBrains</p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </nav>
      <section class="section is-flex-grow-5" id="routed-content">
        <div id="home">
          <div class="container">
            <p>This is the home view...</p>
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a auctor augue. Aenean eget faucibus est. Nullam vestibulum vitae libero at dapibus. Ut urna libero, ultrices eget metus a, pharetra vestibulum lectus. Proin nisi nisi, porttitor sit amet condimentum elementum, scelerisque et nisi. Vestibulum non risus in urna suscipit ornare. Pellentesque lorem ipsum, pulvinar id erat elementum, placerat ultricies massa. Nullam venenatis ex vitae aliquam lacinia. Vestibulum est mi, faucibus ac odio id, dignissim fermentum augue.
Ut vehicula est eget purus laoreet euismod. Morbi eleifend leo eu sapien lacinia dapibus. Suspendisse sed est rhoncus, hendrerit neque quis, euismod magna. Phasellus diam mauris, fringilla eu egestas ut, pellentesque vel odio. Nunc pulvinar nisl in dui auctor interdum. Donec quis erat ac augue imperdiet ultrices. Duis et faucibus quam. Donec convallis dictum volutpat. Vestibulum sed elit sit amet risus sollicitudin posuere pretium vel mauris. Aliquam luctus magna vitae sollicitudin tempus. Maecenas felis orci, mollis et imperdiet vestibulum, lobortis quis elit. Mauris semper quis enim ut vestibulum.
Integer quis est et erat rhoncus tincidunt. Donec eleifend bibendum leo nec rutrum. Nunc maximus est eget eros sodales, sed volutpat magna sodales. Integer fermentum ipsum vel lacus viverra feugiat. Pellentesque ac tincidunt arcu. Mauris elit nunc, posuere congue malesuada a, scelerisque vel sem. Aenean eu est a libero venenatis imperdiet. Nam egestas quam turpis, at eleifend ipsum accumsan vitae. Nullam id mi et lacus scelerisque ornare nec non sem. Nulla vel augue iaculis, blandit mauris pulvinar, mollis massa. Fusce est tortor, commodo quis vestibulum id, pharetra id metus. Nam mi est, laoreet ut pretium eget, efficitur quis nisi. Aenean venenatis sagittis suscipit. Vestibulum a gravida ex.
In malesuada aliquam leo, sit amet pellentesque nibh ultricies vulputate. Aliquam fringilla porta turpis, a congue arcu commodo at. Suspendisse potenti. Integer vel metus lorem. Sed posuere elementum lectus, quis tincidunt nunc vestibulum eu. Pellentesque nec mauris leo. Nullam scelerisque fermentum tellus, vitae hendrerit dolor rutrum et. Cras metus justo, placerat eu eros sit amet, pretium feugiat orci. Aenean ac facilisis odio. Vestibulum non egestas orci. Maecenas rhoncus sem at ornare mattis. Proin ultricies tempus nunc porta fermentum. Nunc varius libero ut elit laoreet efficitur. Nulla lacinia euismod quam eget tristique. Integer sit amet urna id velit venenatis eleifend nec et arcu.
Mauris tincidunt id arcu non euismod. Cras nec eleifend enim. Proin nisi eros, vulputate id elit sed, eleifend dapibus lectus. Donec consectetur ipsum et ex sollicitudin euismod. Etiam varius pharetra feugiat. Aliquam consectetur finibus nunc, at tristique dui ultrices at. Suspendisse potenti. Quisque quis dignissim risus, eu ultrices elit. Nulla convallis commodo risus, a vulputate ligula suscipit vel.
</p>
          </div>
        </div>
      </section>
      <footer class="footer" id="footer">
        <div class="has-text-centered">
          <p><strong>Effit</strong> made with ❤️ by <a href="https://github.com/sch3lp">Sch3lp</a></p>
        </div>
      </footer>
    </div>
  </body>
</html>
